### PR TITLE
fix: more complete parameter info in yaml headers

### LINKED ISSUE #1042: Trestle Catalog Generate does not write parameters For NIST 800-53r4 sample json
## Describe the bug

Per https://github.com/IBM/compliance-trestle/pull/999/files `trestle author catalog-generate` should write the parameters for the control into its markdown file, this works with trestle formatted oscal json:
```
"controls": [
          {
            "id": "ac-1",
            "title": "Access Control Policy and Procedures",
            "params": [
              {
                "id": "ac-1_prm_1",
                "values": [
                  "organization-defined personnel or roles"
                ]
              },
              {
                "id": "ac-1_prm_2",
                "values": [
                  "organization-defined frequency"
                ]
              },
```

but does not work with [the sample file provided by NIST](https://github.com/usnistgov/oscal-content/blob/main/nist.gov/SP800-53/rev4/json/NIST_SP-800-53_rev4_catalog.json):
```
"controls": [
          {
            "id": "ac-1",
            "class": "SP800-53",
            "title": "Access Control Policy and Procedures",
            "params": [
              {
                "id": "ac-1_prm_1",
                "label": "organization-defined personnel or roles"
              },
              {
                "id": "ac-1_prm_2",
                "label": "organization-defined frequency"
              },
              {
                "id": "ac-1_prm_3",
                "label": "organization-defined frequency"
              }
```

## Expected behavior

Parameters in markdown:
```
---
x-trestle-set-params:
  ac-1_prm_1: organization-defined personnel or roles
  ac-1_prm_2: organization-defined frequency
  ac-1_prm_3: organization-defined frequency
---

# ac-1 - \[Access Control\] Access Control Policy and Procedures
```

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation (change which affects the documentation site)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary
More parameter info in yaml headers: label, values, select, how-many, choice
New tag to indicate which parameters should be added as SetParameters during profile-assemble
New allowed-sections option for ssp-generate so that only those sections are included in markdown
Better handling of statement labels: e.g. (a)

closes #1042
## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/common/model_utils.py
Comment (enikonovad) [inline trestle/common/model_utils.py:540]: I think this is checked on lines 536-537?

**Review by enikonovad — APPROVED** [review]
LGTM - small comment and please add docs as discussed before merging.

File: trestle/common/model_utils.py
Comment (fsuits) [reply trestle/common/model_utils.py]: fixed

File: docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md
Comment (enikonovad) [inline docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md:223]: I am missing a part here, what is the benefit of having `profile-values` as separate field? Why can't user directly modify `values`?

File: tests/data/author/controls/control_with_components_and_params.md
Comment (enikonovad) [inline tests/data/author/controls/control_with_components_and_params.md:14]: Are there tests for the headers with list of `values` and choices?

**Review by enikonovad — COMMENTED** [review]
Please see some questions/suggestions. Also was this tested with say `cloud-ssp` repo? Please see some comments below with issues that I have found.

**Comment by enikonovad** [comment]
Tested the branch against `cloud-ssp` repository, and found some issues, not sure it is related to this particular PR. But I think something is broken:

1. Running `trestle author profile-generate --name FSCloud --output FSCloud_Official_guidance -y automation/standard-profile-metadata.yml --set-parameters -ohv` does not create a folder named `FSCloud_Official_guidance`, instead folder named `hv` is created.

2. In the folder `hv` headers have the following structure:
```
x-trestle-set-params:
  ac-1_prm_1: 'Assignment: organization-defined personnel or roles'
  ac-1_prm_2: at least annually
  ac-1_prm_3: at least annually
```

Shouldn’t it be this instead ? :
```
x-trestle-set-params:
  ac-1_prm_1:
    values: 'Assignment: organization-defined personnel or roles'
  ac-1_prm_2:
    values: at least annually
....
```

**Review by enikonovad — CHANGES_REQUESTED** [review]
Please test it against the demo repositories to ensure everything works as intended.

File: docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md
Comment (fsuits) [reply docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md:217]: fixed

File: docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md
Comment (fsuits) [reply docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md:223]: "only those parameter values marked `profile-values` will be part of the assembled profile's SetParams when you assemble with the `--set-parameters` flag."  The "values" field is ignored during assembly because it represents what is coming into the profile.  The `profile-values` entry indicates what - if any - values the profile itself specifies via its SetParameters flag.

It is inherently confusing but I think makes sense once people see it.

**Comment by fsuits** [comment]
All related demos are broken by this change and will need to be updated.  It shouldn't be hard - and they should then point to the new release.  But the -ohv issue is a problem and I will look into.

File: docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md
Comment (enikonovad) [reply docs/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring.md:223]: Okay after reading the code it makes sense to me

**Comment by fsuits** [comment]
I think the issue with -ohv is that you specified --set-parameters when that isn't an option.  It should have error'd out.  I don't think I can duplicate what you did?

File: tests/data/author/controls/control_with_components_and_params.md
Comment (fsuits) [reply tests/data/author/controls/control_with_components_and_params.md:14]: there are tests for the param->dict conversions back and forth including multiple values and validity checks in test_param_to_dict()  I will check if the higher level tests exercise it.

**Review by enikonovad — APPROVED** [review]
LGTM - After discussion the issue was found on my side and the change is working as expected

**Pull request merged**
